### ISSUE TITLE
Make findNiftyControl more typesafe

### ISSUE THREAD

**Issue by toolforger** [nifty-gui/nifty-gui#33]
findNiftyControl cannot currently handle container controls like Menu and DropDown in a typesafe way.
I could write an overloaded variant to deal with that:

findNiftyControl (String elementName, Class<T> containerClazz, Class<E> elementClazz)

Code using the existing, two-parameter findNiftyControl should continue to work, it would simply continue to generate generics warnings.

Useful?

**Comment by void256** [comment]
I don't get it. Can you provide an example please?

**Comment by void256** [comment]
I mean an example for when you would need the new version of findNiftyControl...

**Comment by toolforger** [comment]
In ConsoleControl we have:

```
  private ListBox<String> listBox;
  ...
  this.listBox = element.findNiftyControl("#listBox", ListBox.class);
```

This gives the warning: "Type safety: The expression of type ListBox needs unchecked conversion to conform to ListBox<String>".
The problem is that the compiler knows that findNiftyControl returns a ListBox, but doesn't know the item type.

Basically, I'm trying to eliminate the warning by providing a third Class<?> parameter and telling the compiler to use that as the item type.

**Comment by void256** [comment]
Ah! I see! That's useful and another great idea! :)

We need to make sure to name the new method appropriate and the parameter name: "ElementClazz" is kinda confusing because "element" in Niftys context is already used for the Nifty Element.

What about calling the new method:

```
public < T extends NiftyControl, U > T findNiftyControlOfType(final String elementName, final Class < T > requestedControlClass, final Class < U > requestedControlClassType);
```

Example:

```
ListBox<String> listBox = element.findNiftyControlWithType("#listBox", ListBox.class, String.class);
```

?

PS: We could discuss this directly at irc://freenode/niftygui as well...

**Comment by toolforger** [comment]
Oh, right, "item" would be much better, that's established terminology in ListBox etc.
I'd stick with the original name, findNiftyControl. It's also finding a control of a given type; the new function just uses the item type as an additional parameter.

I.e. the proposal for the signature would be

```
public < T extends NiftyControl, U > T findNiftyControl(
final String elementName, final Class < T > requestedControlClass, final Class < U > requestedItemClass);
```

BTW why are the parameters final? Neither String nor Class are mutable, nor do they allow subclasses that could be.

(I feel awkward in IRC and rarely use it.)

**Comment by void256** [comment]
Since generic methods are sometimes hard to grasp for new people and this method is a central puplic API to access controls I still would like to give an additional hint in the method name when to use the method with one class parameter and where to use the method with two parameters. Just to reduce the possibility of confusion, you know?

But I don't know how to appropriate name the new method :) so I guess keeping the old name and simply adding another parameter is sufficient. Maybe you can add a JavaDoc comment that explains the difference? I think adding the ListBox example from above to the JavaDoc would be great.

The final parameters can't be assigned a new value in the method. Thats all and it works a bit differently in Java as it works in C++ const. In Java the Object represented by the final parameter can still be changed/mutated by calling methods on it. The compiler just makes sure that you can't accidently assign a new value to the final parameter inside the method body - which is usually a bad habit and a source for hard to find errors.

So in my opinion it's a good style to enforce this with making them final. Since at my day job final parameters are enforces via checkstyle I'll add them subconsciously in Nifty as well. But I'm not religious about this :) If you write a new method and you're not using final parameters this would be fine too!

(IRC was just an idea since I had some great discussions with other Nifty users and developers there. Not using it is perfectly fine with me as well :)

**Comment by toolforger** [comment]
I was just curious about the final keyword, I don't mind either way.

Yes, the function will definitely get a good Javadoc.
The signature will indeed be confusing, Java's type erasure forces even more trickery than I thought at first. Heck, I hit a roadblock and thought it wasn't doable at all, but I had a good meal and my subconscious produced an idea that might just work. I'll be trying that in the next hours, so stay tuned :-)

**Comment by toolforger** [comment]
HAH! I got it to work.
Take a look at https://github.com/toolforger/nifty-gui/commit/9e805d9acc9314d0446ea43e87a7825ae5c10a3c .
The new function is in Element.java.
It is used in CreateDropDownControl.java.
To make DropDown usable for the new class, I had to redefine it to inherit from a generic interface. That's NiftyContainerControl<T>, into which I refactored some functions from DropDown (the exact set of functions to move there may not be optimal, I haven't spent much thought on that one yet; this was more a knee-jerk reflex of "oh yeah, if it's going to be a NiftyContainerControl interface then it should define all container-related functions that the subinterfaces share"; maybe this refactoring should go into a separate unit of review).
